src/command_helpers.rs

```diff
@@ -41,83 +40,46 @@ impl CargoOutput {
         }
     }
 
-    pub(crate) fn print_thread(&self) -> Result<Option<PrintThread>, Error> {
-        self.warnings.then(PrintThread::new).transpose()
+    fn stdio_for_warnings(&self) -> Stdio {
+        if self.warnings {
+            Stdio::piped()
+        } else {
+            Stdio::null()
+        }
     }
 }
 
-pub(crate) struct PrintThread {
-    handle: Option<JoinHandle<()>>,
-    pipe_writer: Option<File>,
+pub(crate) fn reader_for_stderr_as_warnings(
+    child: &mut Child,
+) -> Option<BufReader<std::process::ChildStderr>> {
+    child.stderr.take().map(BufReader::new)
```

Comment:
Ok, back to 100b buffer.

The `BufReader` makes things more convenient, otherwise if we were using the `ChildStderr` directly we'd need to shuffle any remaining bytes after the final `\n` back to the beginning of the `Vec` between each `read` OR we would have to `read_to_end` which could result in a very large `Vec`.

Reply:
> otherwise if we were using the ChildStderr directly we'd need to shuffle any remaining bytes after the final \n back 

Hmm I think this isn't very hard with [`Vec::drain`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.drain), just two lines for splitting based on newline and remove it using drain after printing.

Reply:
This feels like a premature optimization: we'd have to choose a size for the `Vec` to read into, which is no different to the buffer in the `BufReader`. It would avoid the second allocation, but this is in the context of spawning and waiting for a process so the execution time is dominated by that rather than the work here. I'd prefer to keep the code simple than make the optimization to avoid a `BufReader`. 

We've already made improvements: 100b for the `BufReader` instead of 4kb and no allocations for the `Vec` if there's no warnings.

Reply:
> It would avoid the second allocation, but this is in the context of spawning and waiting for a process so the execution time is dominated by that rather than the work here. I'd prefer to keep the code simple than make the optimization to avoid a `BufReader`. 

Understood, that is indeed the case.

> We've already made improvements: 100b for the `BufReader` instead of 4kb and no allocations for the `Vec` if there's no warnings. 

Yeah that's probably good enough.

Thanks.